File: test/vectors/cross-emitter-matrix/verify.rb
#!/usr/bin/env ruby
# Cross-emitter matrix — Ruby verifier.

$LOAD_PATH.unshift(File.join(Dir.home, '.gem/ruby/2.6.0/gems/ed25519-1.3.0/lib'))
require 'ed25519'
require 'json'

FIELD_ORDER = %w[
  id timestamp method toolName namespace upstream
  principal durationMs success errorCode previousHash
].freeze

def canonicalize_value(v)
  return nil if v.nil?
  case v
  when String then v
  when TrueClass, FalseClass then v
  when Integer
    raise "unsafe #{v}" if v.abs > (2**53 - 1)
    v
  when Array then ['L', v.map { |x| canonicalize_value(x) }]
  when Hash
    keys = v.keys.sort_by { |k| k.encode('UTF-16BE').bytes }
    ['M', keys.map { |k| [k, canonicalize_value(v[k])] }]
  else raise "unsupported #{v.class}"
  end
end

def canonicalize(record)
  ordered = FIELD_ORDER.map { |k| [k, canonicalize_value(record[k])] }
  %w[decisionContextDigest extensionsDigest aiInvocation parties].each do |opt|
    ordered << [opt, canonicalize_value(record[opt])] unless record[opt].nil?
  end
  JSON.generate(ordered)
end

if ENV['DAEMON_MODE'] == '1'
  # Env pinning: fall back to UTF-8 where the driver did not already set it.
  ENV['LC_ALL'] ||= 'C.UTF-8'
  ENV['PYTHONIOENCODING'] ||= 'utf-8'
  STDIN.set_encoding('UTF-8')
  STDOUT.set_encoding('UTF-8')

  # Reuse VerifyKey instances across records keyed by hex string.
  verify_key_cache = {}

  STDIN.each_line do |line|
    line = line.strip
    next if line.empty?

    # Extract id via regex on the RAW line BEFORE JSON.parse. If the line
    # contains bytes that break JSON.parse (lone surrogates, malformed
    # escapes), req['id'] is unreachable, so id would end up nil and either
    # collide with other nil-id responses or drop entirely. The regex
    # fallback preserves correlation for the driver's
    # response_count == request_count invariant. Note: String#match itself
    # raises on invalid UTF-8 bytes, so scrub a copy first — we leave
    # `line` untouched for JSON.parse so it fails authentically on the bad
    # input.
    id_from_regex = nil
    begin
      safe_line = line.dup.force_encoding('UTF-8').scrub('?')
      id_match = safe_line.match(/"id"\s*:\s*(?:"((?:[^"\\]|\\.)*)"|(-?\d+(?:\.\d+)?))/)
      id_from_regex = id_match ? (id_match[1] || id_match[2]) : nil
    rescue
      id_from_regex = nil
    end

    begin
      req = JSON.parse(line)
      id = req['id']
      payload = req['record']
      pk_hex = payload['public_key_hex']
      verify_key = (verify_key_cache[pk_hex] ||= Ed25519::VerifyKey.new([pk_hex].pack('H*')))

      verified = false
      canonical = nil
      begin
        canonical = canonicalize(payload['record'])
        sig_bytes = [payload['signature_hex']].pack('H*')
        verified = verify_key.verify(sig_bytes, canonical)
      rescue Ed25519::VerifyError
        verified = false
      end

      STDOUT.puts JSON.generate({
        'id' => id,
        'ok' => verified,
        'verified' => verified,
        'local_canonical' => canonical,
        'sig_hex' => payload['signature_hex']
      })
      STDOUT.flush
    rescue Exception => e
      # Paranoid catch: any Exception subclass is normalized into a single
      # well-formed response line so the driver's
      # response_count == request_count invariant holds. The fact-of-
      # rejection (ok:false) is the scientific signal — preserve it; only
      # sanitize the message bytes.
      raw_msg = (e.message rescue 'unknown').to_s
      sanitized_msg = begin
        raw_msg.dup.force_encoding('UTF-8').scrub('?')
      rescue
        begin
          raw_msg.dup.force_encoding('UTF-8')
                 .encode('UTF-8', invalid: :replace, undef: :replace, replace: '?')
        rescue
          'unencodable_error'
        end
      end
      begin
        STDOUT.puts JSON.generate({ 'id' => id_from_regex, 'ok' => false, 'error' => sanitized_msg })
      rescue => _nested
        # Last-resort hand-built JSON so the driver still sees the rejection.
        STDOUT.puts %Q({"id":#{id_from_regex.to_json},"ok":false,"error":"internal_encode_failure"})
      end
      STDOUT.flush
      next
    end
  end
else
  payload = JSON.parse(STDIN.read)
  verify_key = Ed25519::VerifyKey.new([payload['public_key_hex']].pack('H*'))

  verified = false
  canonical = nil
  begin
    canonical = canonicalize(payload['record'])
    sig_bytes = [payload['signature_hex']].pack('H*')
    verified = verify_key.verify(sig_bytes, canonical)
  rescue Ed25519::VerifyError
    verified = false
  rescue => e
    verified = false
    canonical = "ERROR: #{e.message}"
  end

  puts JSON.generate({ 'verified' => verified, 'local_canonical' => canonical, 'sig_hex' => payload['signature_hex'] })
end

File: test/vectors/cross-emitter-matrix/verify.php
#!/usr/bin/env php
<?php
// Cross-emitter matrix — PHP verifier.
const FIELD_ORDER = [
    'id','timestamp','method','toolName','namespace','upstream',
    'principal','durationMs','success','errorCode','previousHash'
];

function canonicalize_value($v) {
    if (is_null($v)) return null;
    if (is_bool($v)) return $v;
    if (is_string($v)) return $v;
    if (is_int($v)) {
        if (abs($v) > (1 << 53) - 1) throw new Exception("unsafe integer $v");
        return $v;
    }
    if (is_float($v)) throw new Exception("unsafe number (float)");
    if (is_array($v)) {
        // JSON arrays only — decoded via stdClass mode so associative arrays
        // never arrive here.
        $inner = [];
        foreach ($v as $vv) {
            $inner[] = canonicalize_value($vv);
        }
        return ['L', $inner];
    }
    if (is_object($v)) {
        // JSON objects arrive as stdClass. foreach over stdClass preserves
        // numeric-string property names as strings (unlike (array)$v or
        // get_object_vars(), both of which coerce "2" back to int 2).
        $keys = [];
        foreach ($v as $k => $vv) {
            $keys[] = $k;
        }
        usort($keys, function($a, $b) {
            $ba = mb_convert_encoding((string)$a, 'UTF-16BE', 'UTF-8');
            $bb = mb_convert_encoding((string)$b, 'UTF-16BE', 'UTF-8');
            return strcmp($ba, $bb);
        });
        $pairs = [];
        foreach ($keys as $k) {
            $pairs[] = [(string)$k, canonicalize_value($v->{$k})];
        }
        return ['M', $pairs];
    }
    throw new Exception("unsupported type");
}

function canonicalize($record) {
    // $record is a stdClass (json_decode without assoc flag).
    $ordered = [];
    foreach (FIELD_ORDER as $k) {
        $v = property_exists($record, $k) ? $record->{$k} : null;
        $ordered[] = [$k, canonicalize_value($v)];
    }
    foreach (['decisionContextDigest', 'extensionsDigest', 'aiInvocation', 'parties'] as $opt) {
        if (property_exists($record, $opt) && $record->{$opt} !== null) {
            $ordered[] = [$opt, canonicalize_value($record->{$opt})];
        }
    }
    return json_encode($ordered, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
}

if (getenv('DAEMON_MODE') === '1') {
    // Env pinning — only set if missing, don't override caller-set values.
    if (getenv('LC_ALL') === false) putenv('LC_ALL=C.UTF-8');
    if (getenv('PYTHONIOENCODING') === false) putenv('PYTHONIOENCODING=utf-8');
    // Cache decoded public keys across records — the reason daemon mode
    // exists is to amortize hex2bin and PHP startup over the batch.
    $pubKeyCache = [];
    while (($line = fgets(STDIN)) !== false) {
        $line = rtrim($line, "\r\n");
        if ($line === '') continue;
        // Best-effort id extraction from the raw line, so we can still emit
        // an {ok:false} correlated response when json_decode itself fails
        // (e.g. lone-surrogate \u escapes in an adversarial record).
        $id = null;
        if (preg_match('/"id"\s*:\s*"((?:\\\\.|[^"\\\\])*)"/', $line, $m)) {
            $decodedId = json_decode('"' . $m[1] . '"');
            $id = ($decodedId === null && json_last_error() !== JSON_ERROR_NONE) ? $m[1] : $decodedId;
        } elseif (preg_match('/"id"\s*:\s*(-?\d+)/', $line, $m)) {
            $id = (int)$m[1];
        } elseif (preg_match('/"id"\s*:\s*(true|false|null)/', $line, $m)) {
            $id = ($m[1] === 'true') ? true : (($m[1] === 'false') ? false : null);
        }
        // Wrap parse + canonicalize + verify in one try so any failure — including
        // json_decode returning null on invalid \u escapes — emits {ok:false}
        // instead of silently dropping the response and breaking id correlation.
        try {
            // stdClass mode preserves numeric-string keys in nested objects.
            $req = json_decode($line);
            if ($req === null && json_last_error() !== JSON_ERROR_NONE) {
                throw new Exception("json_decode: " . json_last_error_msg());
            }
            if (!is_object($req) || !property_exists($req, 'id')) {
                throw new Exception("missing id or non-object request");
            }
            $id = $req->id;
            $pkHex = property_exists($req, 'public_key_hex') ? $req->public_key_hex : null;
            $sigHex = property_exists($req, 'signature_hex') ? $req->signature_hex : null;
            if (!is_string($pkHex) || !is_string($sigHex) || !property_exists($req, 'record')) {
                throw new Exception("missing public_key_hex/signature_hex/record");
            }
            if (!isset($pubKeyCache[$pkHex])) {
                $pubKeyCache[$pkHex] = hex2bin($pkHex);
            }
            $pubKey = $pubKeyCache[$pkHex];
            $sig = hex2bin($sigHex);
            $canonical = canonicalize($req->record);
            $verified = sodium_crypto_sign_verify_detached($sig, $canonical, $pubKey);
            $resp = [
                'id' => $id,
                'ok' => true,
                'verified' => $verified,
                'local_canonical' => $canonical,
                'sig_hex' => $sigHex,
            ];
        } catch (\Throwable $e) {
            if ($id === null) {
                // Truly no id anywhere on the line → cannot correlate; skip.
                continue;
            }
            $resp = ['id' => $id, 'ok' => false, 'error' => $e->getMessage()];
        }
        // JSON_INVALID_UTF8_SUBSTITUTE: raw invalid UTF-8 in the error message
        // (or in the id, when the parent sent an adversarial id) must not cause
        // json_encode to return false — that would drop the rejection response
        // and break the paper's rejection-map signal.
        $out = json_encode($resp, JSON_UNESCAPED_SLASHES | JSON_INVALID_UTF8_SUBSTITUTE);
        if ($out === false) {
            $out = '{"id":' . json_encode((string)$id, JSON_INVALID_UTF8_SUBSTITUTE)
                 . ',"ok":false,"error":"json_encode failed: ' . json_last_error_msg() . '"}';
        }
        fwrite(STDOUT, $out . "\n");
        fflush(STDOUT);
    }
    exit(0);
}

$input = stream_get_contents(STDIN);
// stdClass mode preserves numeric-string keys in nested JSON objects.
$payload = json_decode($input);

$pubKey = hex2bin($payload->public_key_hex);
$sig = hex2bin($payload->signature_hex);

$verified = false;
$canonical = null;
try {
    $canonical = canonicalize($payload->record);
    $verified = sodium_crypto_sign_verify_detached($sig, $canonical, $pubKey);
} catch (Exception $e) {
    $canonical = "ERROR: " . $e->getMessage();
}

echo json_encode([
    'verified' => $verified,
    'local_canonical' => $canonical,
    'sig_hex' => $payload->signature_hex,
]);
